File: backend/src/services/appointmentService.js
const prisma = require('../config/db');
const { logAudit } = require('./auditService');
const { createNotification } = require('./notificationService');
const { addInvoiceLineItem } = require('./billingService');

/**
 * Transactionally completes an outpatient appointment:
 * 1. Sets status to COMPLETED, records diagnosis and prescription.
 * 2. Sets Doctor & StaffProfile availability to AVAILABLE.
 * 3. Idempotently adds Doctor consultation fee to patient Reception invoice.
 * 4. Dispatches LabTest requests if ordered.
 * 5. Notifies Pharmacist if prescription is written.
 * 6. Logs Audit record.
 */
async function completeAppointmentTransaction({
  appointmentId,
  diagnosis = null,
  prescription = null,
  orderedTests = [],
  completedByUserId = null,
  ipAddress = null
}) {
  return await prisma.$transaction(async (tx) => {
    const appointment = await tx.appointment.findUnique({
      where: { id: appointmentId },
      include: {
        doctor: { include: { user: true } },
        patient: true
      }
    });

    if (!appointment) {
      throw new Error(`Appointment ${appointmentId} not found`);
    }

    if (appointment.status === 'COMPLETED') {
      return { appointment, alreadyCompleted: true };
    }

    // 1. Update appointment
    const updatedAppointment = await tx.appointment.update({
      where: { id: appointmentId },
      data: {
        status: 'COMPLETED',
        diagnosis: diagnosis || appointment.diagnosis,
        prescription: prescription || appointment.prescription
      },
      include: { doctor: { include: { user: true } }, patient: true }
    });

    // 2. Set Doctor and StaffProfile availability back to AVAILABLE
    if (appointment.doctorId) {
      await tx.doctor.update({
        where: { id: appointment.doctorId },
        data: { availability: 'AVAILABLE' }
      });

      if (appointment.doctor?.userId) {
        await tx.staffProfile.updateMany({
          where: { userId: appointment.doctor.userId },
          data: { availability: 'AVAILABLE' }
        });
      }
    }

    // 3. Automated Consultation Fee Billing (Idempotent)
    const consultationFee = appointment.doctor.consultationFee || 100.00;
    const docName = appointment.doctor.user?.name || 'Attending Physician';

    await addInvoiceLineItem({
      patientId: appointment.patientId,
      billingType: 'RECEPTION',
      sourceDepartment: 'CLINICAL',
      itemBillingType: 'CONSULTATION',
      itemDescription: `Doctor Consultation - ${docName} (${appointment.doctor.specialization})`,
      quantity: 1,
      unitPrice: consultationFee,
      sourceEntity: 'APPOINTMENT',
      sourceId: appointment.id,
      prismaClient: tx
    });

    // 4. Create Lab Tests if ordered during consultation
    const createdLabTests = [];
    if (Array.isArray(orderedTests) && orderedTests.length > 0) {
      for (const t of orderedTests) {
        if (t.testName) {
          const test = await tx.labTest.create({
            data: {
              patientId: appointment.patientId,
              testName: t.testName,
              category: t.category || 'Diagnostics',
              cost: t.cost ? parseFloat(t.cost) : 50.00,
              requestedBy: docName,
              status: 'PENDING'
            }
          });
          createdLabTests.push(test);

          // Alert Lab Tech
          await createNotification({
            role: 'LAB_TECHNICIAN',
            title: 'New Diagnostic Test Ordered',
            message: `${t.testName} ordered for patient ${appointment.patient.firstName} ${appointment.patient.lastName} by ${docName}.`,
            type: 'LAB_REQUEST',
            entityId: test.id,
            prismaClient: tx
          });
        }
      }
    }

    // 5. Notify Pharmacist if prescription given
    if (prescription && prescription.trim().length > 0) {
      await createNotification({
        role: 'PHARMACIST',
        title: 'New Patient Prescription',
        message: `Prescription issued for patient ${appointment.patient.firstName} ${appointment.patient.lastName} (${appointment.patient.mrn}) by ${docName}.`,
        type: 'GENERAL',
        entityId: appointment.id,
        prismaClient: tx
      });
    }

    // 6. Audit logging
    await logAudit({
      userId: completedByUserId,
      action: 'APPOINTMENT_COMPLETED',
      resource: 'APPOINTMENT',
      details: {
        appointmentId: appointment.id,
        patientMrn: appointment.patient.mrn,
        doctorName: docName,
        consultationFee,
        testsOrderedCount: createdLabTests.length
      },
      ipAddress,
      prismaClient: tx
    });

    return {
      appointment: updatedAppointment,
      createdLabTests,
      alreadyCompleted: false
    };
  });
}

module.exports = {
  completeAppointmentTransaction
};

File: backend/app/services/appointment_service.py
from typing import Optional, List, Dict, Any
from sqlalchemy.orm import Session
from app import models
from app.services.billing_service import add_invoice_line_item
from app.services.notification_service import create_notification
from app.services.audit_service import log_audit

def complete_appointment_transaction(
    db: Session,
    appointment_id: str,
    diagnosis: Optional[str] = None,
    prescription: Optional[str] = None,
    ordered_tests: Optional[List[Dict[str, Any]]] = None,
    completed_by_user_id: Optional[str] = None,
    ip_address: Optional[str] = None
) -> Dict[str, Any]:
    appointment = db.query(models.Appointment).filter(models.Appointment.id == appointment_id).first()
    if not appointment:
        raise ValueError(f"Appointment {appointment_id} not found")

    if appointment.status == "COMPLETED":
        return {"appointment": appointment, "alreadyCompleted": True, "createdLabTests": []}

    # 1. Update appointment
    appointment.status = "COMPLETED"
    if diagnosis:
        appointment.diagnosis = diagnosis
    if prescription:
        appointment.prescription = prescription

    # 2. Reset Doctor & StaffProfile availability
    if appointment.doctorId and appointment.doctor:
        appointment.doctor.availability = "AVAILABLE"
        if appointment.doctor.userId:
            db.query(models.StaffProfile).filter(models.StaffProfile.userId == appointment.doctor.userId).update({"availability": "AVAILABLE"})

    # 3. Add Consultation Fee Line Item
    consultation_fee = appointment.doctor.consultationFee if appointment.doctor else 100.00
    doc_name = appointment.doctor.user.name if (appointment.doctor and appointment.doctor.user) else "Attending Physician"
    doc_spec = appointment.doctor.specialization if appointment.doctor else "General"

    add_invoice_line_item(
        db=db,
        patient_id=appointment.patientId,
        billing_type="RECEPTION",
        source_department="CLINICAL",
        item_billing_type="CONSULTATION",
        item_description=f"Doctor Consultation - {doc_name} ({doc_spec})",
        quantity=1,
        unit_price=consultation_fee,
        source_entity="APPOINTMENT",
        source_id=appointment.id
    )

    # 4. Create Lab Tests if ordered
    created_lab_tests = []
    if ordered_tests:
        for t in ordered_tests:
            test_name = t.get("testName") if isinstance(t, dict) else getattr(t, "testName", None)
            if test_name:
                category = t.get("category", "Diagnostics") if isinstance(t, dict) else getattr(t, "category", "Diagnostics")
                cost = float(t.get("cost", 50.0)) if isinstance(t, dict) else float(getattr(t, "cost", 50.0))

                lab_test = models.LabTest(
                    patientId=appointment.patientId,
                    testName=test_name,
                    category=category,
                    cost=cost,
                    requestedBy=doc_name,
                    status="PENDING"
                )
                db.add(lab_test)
                db.flush()
                created_lab_tests.append(lab_test)

                pat_name = f"{appointment.patient.firstName} {appointment.patient.lastName}" if appointment.patient else "Patient"
                create_notification(
                    db=db,
                    role="LAB_TECHNICIAN",
                    title="New Diagnostic Test Ordered",
                    message=f"{test_name} ordered for patient {pat_name} by {doc_name}.",
                    type="LAB_REQUEST",
                    entity_id=lab_test.id
                )

    # 5. Notify Pharmacist if prescription written
    if prescription and prescription.strip():
        pat_name = f"{appointment.patient.firstName} {appointment.patient.lastName}" if appointment.patient else "Patient"
        pat_mrn = appointment.patient.mrn if appointment.patient else "N/A"
        create_notification(
            db=db,
            role="PHARMACIST",
            title="New Patient Prescription",
            message=f"Prescription issued for patient {pat_name} ({pat_mrn}) by {doc_name}.",
            type="GENERAL",
            entity_id=appointment.id
        )

    # 6. Audit logging
    pat_mrn = appointment.patient.mrn if appointment.patient else "N/A"
    log_audit(
        db=db,
        user_id=completed_by_user_id,
        action="APPOINTMENT_COMPLETED",
        resource="APPOINTMENT",
        details={
            "appointmentId": appointment.id,
            "patientMrn": pat_mrn,
            "doctorName": doc_name,
            "consultationFee": consultation_fee,
            "testsOrderedCount": len(created_lab_tests)
        },
        ip_address=ip_address
    )

    db.commit()
    return {
        "appointment": appointment,
        "createdLabTests": created_lab_tests,
        "alreadyCompleted": False
    }
